packages/react-core/src/helpers/util.ts

```diff
@@ -273,3 +273,101 @@ export const toCamel = (s: string) => s.replace(/([-_][a-z])/gi, camelize);
+
+/**
+ * Calculate the width of the text
+ * Example:
+ * getTextWidth('my text', node)
+ *
+ * @param {string} text The text to calculate the width for
+ * @param {HTMLElement} node The HTML element
+ */
+export const getTextWidth = (text: string, node: HTMLElement) => {
+  const computedStyle = getComputedStyle(node);
+  // Firefox returns the empty string for .font, so this function creates the .font property manually
+  const getFontFromComputedStyle = () => {
+    let computedFont = '';
+    // Firefox uses percentages for font-stretch, but Canvas does not accept percentages
+    // so convert to keywords, as listed at:
+    // https://developer.mozilla.org/en-US/docs/Web/CSS/font-stretch
+    const fontStretchLookupTable = {
+      '50%': 'ultra-condensed',
+      '62.5%': 'extra-condensed',
+      '75%': 'condensed',
+      '87.5%': 'semi-condensed',
+      '100%': 'normal',
+      '112.5%': 'semi-expanded',
+      '125%': 'expanded',
+      '150%': 'extra-expanded',
+      '200%': 'ultra-expanded'
+    };
+    // If the retrieved font-stretch percentage isn't found in the lookup table, use
+    // 'normal' as a last resort.
+    let fontStretch;
+    if (computedStyle.fontStretch in fontStretchLookupTable) {
+      fontStretch = (fontStretchLookupTable as any)[computedStyle.fontStretch];
+    } else {
+      fontStretch = 'normal';
+    }
+    computedFont =
+      computedStyle.fontStyle +
+      ' ' +
+      computedStyle.fontVariant +
+      ' ' +
+      computedStyle.fontWeight +
+      ' ' +
+      fontStretch +
+      ' ' +
+      computedStyle.fontSize +
+      '/' +
+      computedStyle.lineHeight +
+      ' ' +
+      computedStyle.fontFamily;
+    return computedFont;
+  };
+
+  const canvas = document.createElement('canvas');
+  const context = canvas.getContext('2d');
+  context.font = computedStyle.font || getFontFromComputedStyle();
+
+  return context.measureText(text).width;
+};
+
+/**
+ * Get the inner dimensions of an element
+ *
+ * @param {HTMLElement} node HTML element to calculate the inner dimensions for
+ */
+export const innerDimensions = (node: HTMLElement) => {
+  const computedStyle = getComputedStyle(node);
+
+  let width = node.clientWidth; // width with padding
+  let height = node.clientHeight; // height with padding
+
+  height -= parseFloat(computedStyle.paddingTop) + parseFloat(computedStyle.paddingBottom);
+  width -= parseFloat(computedStyle.paddingLeft) + parseFloat(computedStyle.paddingRight);
+  return { height, width };
+};
+
+/**
+ * This function is a helper for truncating text content on the left, leaving the right side of the content in view
+ *
+ * @param {HTMLElement} node HTML element
+ * @param {string} value The original text value
+ */
+export const trimLeft = (node: HTMLElement, value: string) => {
+  const availableWidth = innerDimensions(node).width;
+  let newValue = value;
+  if (getTextWidth(value, node) > availableWidth) {
```

Comment:
just one bug here, can you add an `else` condition to this `if`. Noticed that if you just a little overflow and you expand the textbox until the whole text should fit again, it won't restore the value
```
 else {
    if ((node as HTMLInputElement).value) {
      (node as HTMLInputElement).value = value;
    } else {
      node.innerText = value;
    }
  }
```